pkg/exchange/coinbase/stream_handlers.go

```diff
@@ -0,0 +1,344 @@
+package coinbase
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/types"
+)
+
+type channelType struct {
+	Name       string   `json:"name"`
+	ProductIDs []string `json:"product_ids,omitempty"`
+}
+
+type authMsg struct {
+	Signature  string `json:"signature,omitempty"`
+	Key        string `json:"key,omitempty"`
+	Passphrase string `json:"passphrase,omitempty"`
+	Timestamp  string `json:"timestamp,omitempty"`
+}
+
+type subscribeMsgType1 struct {
+	Type     string        `json:"type"`
+	Channels []channelType `json:"channels"`
+
+	authMsg
+}
+
+type subscribeMsgType2 struct {
+	Type       string   `json:"type"`
+	Channels   []string `json:"channels"`
+	ProductIDs []string `json:"product_ids,omitempty"`
+	AccountIDs []string `json:"account_ids,omitempty"` // for balance channel
+
+	authMsg
+}
+
+func (s *Stream) handleConnect() {
+	// subscribe to channels
+	if len(s.Subscriptions) == 0 {
+		return
+	}
+
+	subProductsMap := make(map[string][]string)
+	for _, sub := range s.Subscriptions {
+		strChannel := string(sub.Channel)
+		// rfqMatchChannel allow empty symbol
+		if sub.Channel != rfqMatchChannel && len(sub.Symbol) == 0 {
+			continue
+		}
+		subProductsMap[strChannel] = append(subProductsMap[strChannel], sub.Symbol)
+	}
+	var subCmds []any
+	signature, ts := s.generateSignature()
+	for channel, productIDs := range subProductsMap {
+		var subType string
+		switch channel {
+		case "rfq_matches":
+			subType = "subscriptions"
+		default:
+			subType = "subscribe"
+		}
+		var subCmd any
+		switch channel {
+		case "ticker", "ticker_batch", "level2", "level2_batch":
+			subCmd = subscribeMsgType2{
+				Type:       subType,
+				Channels:   []string{channel},
+				ProductIDs: productIDs,
+				authMsg: authMsg{
+					Signature:  signature,
+					Key:        s.apiKey,
+					Passphrase: s.passphrase,
+					Timestamp:  ts,
+				},
+			}
+		case "balance":
+			subCmd = subscribeMsgType2{
+				Type:       subType,
+				Channels:   []string{channel},
+				AccountIDs: productIDs,
+
+				authMsg: authMsg{
+					Signature:  signature,
+					Key:        s.apiKey,
+					Passphrase: s.passphrase,
+					Timestamp:  ts,
+				},
+			}
+		default:
+			subCmd = subscribeMsgType1{
+				Type: subType,
+				Channels: []channelType{
+					{
+						Name:       channel,
+						ProductIDs: productIDs,
+					},
+				},
+
+				authMsg: authMsg{
+					Signature:  signature,
+					Key:        s.apiKey,
+					Passphrase: s.passphrase,
+					Timestamp:  ts,
+				},
+			}
+		}
+		subCmds = append(subCmds, subCmd)
+	}
+	for _, subCmd := range subCmds {
+		err := s.Conn.WriteJSON(subCmd)
+		if err != nil {
+			log.WithError(err).Errorf("subscription error: %v", subCmd)
```

Comment:
Long with %v prints all fields of the struct. The struct contains exchange credentials. We shouldn't log the credentials. Note that error logs may also be sent to 3 party systems such as rollbar or sentry. It leaks the credential.

Reply:
Fixed.
I also add a test on it.